nimbus/src/enrollment.rs

```diff
@@ -85,90 +365,156 @@ pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
+        let updated_experiments = map_experiments(&updated_experiments);
+        let existing_enrollments = map_enrollments(&existing_enrollments);
+
+        let mut all_slugs = HashSet::with_capacity(existing_experiments.len());
+        all_slugs.extend(existing_experiments.keys());
+        all_slugs.extend(updated_experiments.keys());
+        all_slugs.extend(existing_enrollments.keys());
+
+        for slug in all_slugs {
+            let updated_enrollment = self.evolve_enrollment(
+                is_user_participating,
+                existing_experiments.get(slug).copied(),
+                updated_experiments.get(slug).copied(),
+                existing_enrollments.get(slug).copied(),
+                &mut enrollment_events,
+            )?;
+            if let Some(enrollment) = updated_enrollment {
+                updated_enrollments.push(enrollment);
+            }
+        }
+
+        Ok((updated_enrollments, enrollment_events))
+    }
+
+    fn evolve_enrollment(
+        &self,
+        is_user_participating: bool,
+        existing_experiment: Option<&Experiment>,
+        updated_experiment: Option<&Experiment>,
+        existing_enrollment: Option<&ExperimentEnrollment>,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Option<ExperimentEnrollment>> {
+        Ok(
+            match (existing_experiment, updated_experiment, existing_enrollment) {
+                // New experiment.
+                (None, Some(experiment), None) => Some(ExperimentEnrollment::from_new_experiment(
+                    is_user_participating,
+                    self.nimbus_id,
+                    self.available_randomization_units,
+                    self.app_context,
+                    experiment,
+                    enrollment_events,
+                )?),
+                // Experiment deleted remotely.
+                (Some(_), None, Some(enrollment)) => {
+                    enrollment.on_experiment_discarded(enrollment_events)
+                }
+                // Known experiment.
+                (Some(_), Some(updated_experiment), Some(existing_enrollment)) => {
+                    existing_enrollment.on_experiment_updated(
+                        is_user_participating,
+                        self.nimbus_id,
+                        self.available_randomization_units,
+                        self.app_context,
+                        updated_experiment,
+                        enrollment_events,
+                    )?
+                }
+                (None, None, Some(_)) => unreachable!("Orphan enrollment."),
+                (None, Some(_), Some(_)) => {
+                    unreachable!("New experiment but enrollment already exists.")
+                }
+                (Some(_), None, None) | (Some(_), Some(_), None) => {
+                    unreachable!("Every experiment in the DB has an associated enrollment record.")
+                }
```

Comment:
These are important invariants which feel like they should be documented somewhere, perhaps in the persistence layer schema documentation:

* Every known experiment should have a recorded enrollment.
* Enrollments records for unknown experiments should only ever be `WasEnrolled`.

I also wonder whether it's worth adding some asserts to enforce these, e.g. you could do a consistency check at the end of `evolve_enrollments_in_db` to make sure we're not putting the db into a bad state.

(An alternative way of enforcing these invariants, would be to store the experiment info as a field on the enrollment record, so that you can never store one without its corresponding other; not sure that would be worth the storage schema migration though)